bitnami/mongodb/templates/secrets-ca.yaml

```diff
@@ -0,0 +1,21 @@
+{{- if .Values.tls.enabled -}}
+{{- $cn := printf "%s-headless.%s.svc.cluster.local" ( include "mongodb.fullname" . ) .Release.Namespace }}
```

Comment:
So the headless piece can be deleted. I think this was a mistype. I tested it without the headless to simple be 
*{{- $cn := printf "%s.%s.svc.cluster.local" ( include "mongodb.fullname" . ) .Release.Namespace }}*
and it works fine. On the default non TLS bitnami/mongodb chart, when you run it you get the following:

Reply:
So the headless piece can be deleted. I think this was a mistype. I tested it without the headless to simple be 
*{{- $cn := printf "%s.%s.svc.cluster.local" ( include "mongodb.fullname" . ) .Release.Namespace }}*
and it works fine. On the default non TLS bitnami/mongodb chart, when you run it you get the following:
MongoDB can be accessed via port 27017 on the following DNS name(s) from within your cluster:

    test-mongodb-0.test-mongodb-headless.default.svc.cluster.local
    test-mongodb-1.test-mongodb-headless.default.svc.cluster.local
    test-mongodb-2.test-mongodb-headless.default.svc.cluster.local
    test-mongodb-3.test-mongodb-headless.default.svc.cluster.local

When you exec into the container you also get : 
###
I have no name!@test-mongodb-0:/$ env
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_SERVICE_PORT=443
HOSTNAME=test-mongodb-0
MY_POD_NAMESPACE=default
K8S_SERVICE_NAME=test-mongodb-headless
MONGODB_REPLICA_SET_NAME=rs0
MONGODB_ADVERTISED_HOSTNAME=test-mongodb-0.test-mongodb-headless.default.svc.cluster.local
MY_POD_NAME=test-mongodb-0
MONGODB_INITIAL_PRIMARY_HOST=test-mongodb-0.test-mongodb-headless.default.svc.cluster.local

I have no name!@test-mongodb-0:/$ hostname -A
test-mongodb-0.test-mongodb-headless.default.svc.cluster.local 
###

Therefore I think there was a misconception of the need to include *headless* in that string because of this. However that is not the case. We are generating the certs for the mongodb.fullname. 

When you look at the statefulset.yaml file, in the initContainer **name: generate-tls-certs** , you will notice there is a **svc** field. If you do not add the **svc=** in the statefulset.yaml file, it causes the generated server cert to not include the *headless* in the CN therefore it would not match the remote host which is "test-mongodb-0.test-mongodb-headless.default.svc.cluster.local
see below the logs when the initContainer does not include the svc:
{"t":{"$date":"2020-10-28T12:32:33.268+00:00"},"s":"E",  "c":"NETWORK",  "id":23257,   "ctx":"ReplicaSetMonitor-TaskExecutor","msg":"The server certificate does not match the remote host name","attr":{"remoteHost":"test-mongodb-0.test-mongodb-headless.default.svc.cluster.local","certificateNames":"SAN(s): test-mongodb-0, test-mongodb-0, test-mongodb-0..default.svc.cluster.local, localhost, 127.0.0.1, CN: test-mongodb-0"}}

Looking at the generated openssl.cnf file in the pod confirms this as well as you will see its missing the **headless** in the SAN, thats why it wont connect.
DNS.3 = test-mongodb-0..default.svc.cluster.local

HOwever once that svc is included, it adds the headless in the full CN and allows the server cert to match the remote hosrt